### PR TITLE
Add UpdateHub agent HTTP API support

### PR DESCRIPTION
**Pull request by Jonathas-Conceicao**
- API Blueprint containing Report action used to inform server of client's current state and error messages.

### CODE REVIEW

**Comment by otavio** [comment]
Hello @Jonathas-Conceicao,

Indeed, we ought to add the doc here and start to implement the routes. 

The following would be the needed:

- [ ] Add documentation of API
- [ ] Add `/info`
- [ ] Add `/probe`
- [ ] Add `/update/download/abort`
- [ ] Add `/report`
- [ ] Add `/log`

File: doc/agent-http.apib [hunk 1]
```diff
@@ -0,0 +1,228 @@
+FORMAT: 1A
```
Comment (otavio) [inline doc/agent-http.apib]: The commit log might be improved, something on line of:

```
doc: Add HTTP API documentation

This adds the HTTP API documentation using the API Blueprint
format.

Signed-off-by...
```

File: src/agent.rs [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    let agent_clone = agent.clone();
+    let abort = path!("update" / "download" / "abort").map(move || agent_clone.abort());
+
+    let gets = warp::get2().and(info.or(log));
+    let posts = warp::post2().and(probe.or(abort));
+    gets.or(posts).boxed()
+}
+
+#[derive(Clone)]
+struct Agent {
```
Comment (otavio) [inline src/agent.rs]: The general idea seems very good and I like the approach being used, however, when I read Agent it does not communicate what it is intending to provide. 

Please consider naming the modulle `api` and the struct `Api` or `Handler`. I might be missing something so please say so if it is the case.

File: src/agent_http.rs [hunk 3]
```diff
@@ -0,0 +1,107 @@
+// Copyright (C) 2019 O.S. Systems Sofware LTDA
+//
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{firmware::Metadata, settings::Settings, mem_drain::MemDrain};
+
+use serde_derive::Serialize;
```
Comment (otavio) [inline src/agent_http.rs]: You can use:

```
use serde::Serialize;
```

File: src/agent_http.rs [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+
+    let agent_clone = agent.clone();
+    let info = path!("info").map(move || agent_clone.info());
+    let log = path!("log").map(API::log);
+
+    let agent_clone = agent.clone();
+    let probe = path!("probe").map(move || agent_clone.probe(None));
+    let agent_clone = agent.clone();
+    let probe = probe.or(
+        warp::path::param::<String>().map(move |server_address| agent_clone.probe(Some(server_address))),
```
Comment (otavio) [inline src/agent_http.rs]: Did you try using `warp::path::param::<Option<String>>()`? This is untested but would make sense to just work.

File: src/agent_http.rs [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+            firmware: &'a Metadata,
+        }
+
+        let payload = Payload {
+            version: crate::version(),
+            config: &self.settings,
+            firmware: &self.firmware,
+        };
+
+        if let Ok(response) = serde_json::to_string_pretty(&payload) {
```
Comment (otavio) [inline src/agent_http.rs]: I think warp knows how to serialize JSON if it is the format. So I believe you could just return the warp reply here inside of a `Result`. Does it makes sense?

File: src/agent_http.rs [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+            return response;
+        }
+        panic!("Failed to serialize payload for /info");
```
Comment (otavio) [inline src/agent_http.rs]: We cannot panic, this is an embedded update agent so we must to be as safe as possible. See if we couldn't use the `warp::reject` or similar here.

File: src/agent_http.rs (hunk 3 above)
Comment (Jonathas-Conceicao) [reply src/agent_http.rs]: I need the `serde_derive::Serialize` because it is used for the derive Serialize macro.

File: src/agent_http.rs (hunk 4 above)
Comment (Jonathas-Conceicao) [reply src/agent_http.rs]: The compiler complains that:
``` the trait `std::str::FromStr` is not implemented for `std::option::Option<std::string::String>` ```
I could create an internal type, say `StringOption`, that implements `FromStr` and gives `None` for empty strings. I think this would work as expected.

File: src/agent_http.rs (hunk 5 above)
Comment (Jonathas-Conceicao) [reply src/agent_http.rs]: To that the compiler gives me:
``` the trait `warp::reply::sealed::ReplySealed` is not implemented for `impl firmware::_IMPL_SERIALIZE_FOR_Metadata::_serde::Serialize` ```
On the `BoxedFilter` returned by `routes`. So I'm sticking to the current explicit conversion to JSON for now.

File: src/agent_http.rs (hunk 6 above)
Comment (Jonathas-Conceicao) [reply src/agent_http.rs]: I'm setting the responses to the type `warp::reply::WithStatus` and replying with INTERNAL_SERVER_ERROR and a String with the message instead of panicking.

File: src/agent_http.rs (hunk 3 above)
Comment (otavio) [reply src/agent_http.rs]: Add `derive` to `serde` features inside the `Cargo.toml` and it hides it for you.

File: src/agent_http.rs (hunk 4 above)
Comment (otavio) [reply src/agent_http.rs]: Too much work. It is fine as it is.

File: src/states/mod.rs [hunk 7]
```diff
@@ -25,12 +26,11 @@ use slog_scope::debug;
...
 trait StateChangeImpl {
     fn handle(self) -> Result<StateMachine, failure::Error>;
-}
-
-trait TransitionCallback: StateChangeImpl + Into<State<Idle>> {
     fn callback_state_name(&self) -> &'static str;
 }
 
+trait TransitionCallback: StateChangeImpl + Into<State<Idle>> {}
+
```
Comment (otavio) [inline src/states/mod.rs]: This should be another commit.

File: src/settings.rs [hunk 8]
```diff
@@ -2,11 +2,11 @@
 //
 // SPDX-License-Identifier: Apache-2.0
 
-use crate::serde_helpers::de;
+use crate::serde_helpers::{de, ser};
 
 use chrono::Duration;
 use failure::Fail;
-use serde_derive::Deserialize;
+use serde_derive::{Deserialize, Serialize};
```
Comment (otavio) [inline src/settings.rs]: Avoid `serde_derive` use adding the `derive` feature for `serde`

File: doc/agent-http.apib [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
```
Comment (otavio) [inline doc/agent-http.apib]: This commit can be squashed onto the one which adds the API support.

File: src/mem_drain.rs [hunk 10]
```diff
@@ -41,9 +41,9 @@ impl ToString for MemDrain {
 
         let mut ret = String::new();
         for record in records.iter() {
-            write!(
+            writeln!(
                 &mut ret,
-                "{} {} {}\n",
+                "{} {} {}",
```
Comment (otavio) [inline src/mem_drain.rs]: This is part of the previous mem_drain rework.
